common/coin_support/eth.c

```diff
@@ -271,59 +271,60 @@ uint64_t hex2dec(const char *source) {
-{
-    eth_utxn_ptr->contract_verified = 1;
-    if (eth_utxn_ptr->chain_id_size[0] == 0 || eth_utxn_ptr->nonce_size[0] == 0)
-        return false;
-    if (is_zero(eth_utxn_ptr->value, eth_utxn_ptr->value_size[0]) && eth_utxn_ptr->payload_size == 0)
-        return false;
-    if (is_zero(eth_utxn_ptr->gas_limit, eth_utxn_ptr->gas_limit_size[0]))
-        return false;
-    if (is_zero(eth_utxn_ptr->gas_price, eth_utxn_ptr->gas_price_size[0]))
-        return false;
-    if (cy_read_be(eth_utxn_ptr->chain_id, eth_utxn_ptr->chain_id_size[0]) != metadata_ptr->network_chain_id)
-        return false;
-    // do not allow transaction with unknown payload
-    if (eth_utxn_ptr->payload_size > 0 && U32_READ_BE_ARRAY(eth_utxn_ptr->payload) != TRANSFER_FUNC_SIGNATURE)
-        return false;
+bool eth_validate_unsigned_txn(const eth_unsigned_txn *eth_utxn_ptr, txn_metadata *metadata_ptr) {
+    return !((eth_utxn_ptr->chain_id_size[0] == 0 || eth_utxn_ptr->nonce_size[0] == 0) ||
+             (is_zero(eth_utxn_ptr->value, eth_utxn_ptr->value_size[0]) && eth_utxn_ptr->payload_size == 0) ||
+             (is_zero(eth_utxn_ptr->gas_limit, eth_utxn_ptr->gas_limit_size[0])) ||
+             (is_zero(eth_utxn_ptr->gas_price, eth_utxn_ptr->gas_price_size[0])) ||
+             (cy_read_be(eth_utxn_ptr->chain_id, eth_utxn_ptr->chain_id_size[0]) != metadata_ptr->network_chain_id) ||
+             (eth_utxn_ptr->payload_status == PAYLOAD_CONTRACT_INVALID));
+}
 
-    // Check if payload exists, it's a token transfer operation on EVM
-    if (eth_utxn_ptr->payload_size > 0 && U32_READ_BE_ARRAY(eth_utxn_ptr->payload) == TRANSFER_FUNC_SIGNATURE) {
-        eth_utxn_ptr->contract_verified = 0;
+static PAYLOAD_STATUS eth_decode_txn_payload(const eth_unsigned_txn *eth_utxn_ptr, const txn_metadata *metadata_ptr) {
+  PAYLOAD_STATUS result = PAYLOAD_ABSENT;
+  if (eth_utxn_ptr->payload_size > 0) {
+    switch (U32_READ_BE_ARRAY(eth_utxn_ptr->payload)) {
+      case TRANSFER_FUNC_SIGNATURE:
+        result = PAYLOAD_CONTRACT_NOT_WHITELISTED;
         for (int16_t i = 0; i < WHITELISTED_CONTRACTS_COUNT; i++) {
-            // Verify the contract addresses of whitelisted token symbol
-            if (strncmp(metadata_ptr->token_name, whitelisted_contracts[i].symbol, ETHEREUM_TOKEN_SYMBOL_LENGTH) == 0) {
-                eth_utxn_ptr->contract_verified = 1;
-                return (memcmp(eth_utxn_ptr->to_address, whitelisted_contracts[i].address, ETHEREUM_ADDRESS_LENGTH) == 0);
-            }
+          if (strncmp(metadata_ptr->token_name, whitelisted_contracts[i].symbol, ETHEREUM_TOKEN_SYMBOL_LENGTH) == 0) {
+            result = (memcmp(eth_utxn_ptr->to_address, whitelisted_contracts[i].address, ETHEREUM_ADDRESS_LENGTH) == 0)
+                         ? PAYLOAD_WHITELISTED
+                         : PAYLOAD_CONTRACT_INVALID;
+            break;
+          }
         }
+        break;
+      default:
+        result = PAYLOAD_SIGNATURE_NOT_WHITELISTED;
```

Comment:
add a todo here to whitelist supported functions for clear signing.

otherwise, this will override all supported functions enabled for clear signing and make them not whitelisted.

essentially, this disables clear signing. 
```suggestion
        // TODO: whitelist functions enabled for clear siginng
        result = PAYLOAD_SIGNATURE_NOT_WHITELISTED;
```